### ISSUE TITLE
mocha-allure-reporter trouble: Cannot read property 'currentTest' of undefined

### ISSUE THREAD

**Issue by alexKazarin** [yandex/mocha-parallel-tests#80]
Trying to use mocha-parallel-tests  for selenium-webdriver UI-tests and with _--reporter mocha-allure-reporter_, but got a trouble.
Simple example:

```
var tests = require('selenium-webdriver/testing');
var webdriver = require('selenium-webdriver');

tests.describe('Array', function() {
    tests.it('should return -1 when the value is not present', function(done) {
      var browser = new webdriver.Builder().usingServer('')
          .withCapabilities({
            'browserName': 'chrome',
            'timeout': '3000'
          }).build();
      browser.get('http://www.ya.ru');
      browser.quit();
      allure.feature('featureName');
      done();
    });
});
```

When execute: 

>  mocha --reporter mocha-allure-reporter ./src/simpletest.js  

webdriver visually starts, open page, quit and successfully generate report with "label name="feature" value="featureName"" as it must be with use of _allure.feature('featureName');_

But when I try:

>  ./node_modules/.bin/mocha-parallel-tests --reporter mocha-allure-reporter ./src/simpletest.js

webdriver doesn't even start and in report-file I've got: _Cannot read property 'currentTest' of undefined_

My versions:
- node v6.8.0
- mocha v2.4.5
- mocha-parallel-tests v1.2.2
- mocha-allure-reporter v1.2.3
- selenium-webdriver v2.53.1

Thank you in advance!

**Comment by 1999** [comment]
Thank you for posting issue. From what I see this issue is related to #79 and should be fixed when we merge this PR. Unfortunately there are memory leaks there and PR seems to last a bit longer than I expected, but I hope I will finish it by Monday.

**Comment by alexKazarin** [comment]
Have just tested new variant, unfortunately the same problem. Use console.log in test:
```
console.log(allure._allure);
```
When use with mocha:
```
Allure {
  suites: 
   [ Suite {
       name: 'Array',
       start: 1477299247037,
       testcases: [Object],
       currentTest: [Object],
       currentStep: [Object] },
     Suite { name: '', start: 1477299247036, testcases: [] } ],
  options: { targetDir: 'allure-results' } }
```
But when try to use with mocha-parallel-tests - it's empty:
```
Allure { suites: [], options: { targetDir: 'allure-results' } }
```
That's why allure could not be used in test like this: *allure.feature('featureName');*

**Comment by 1999** [comment]
The problem seems to be interesting. This is the debug from allure-commons:
```
Event occured: [suite]
[1477435036533] startSuite call (suites number: 1, id: 0.689213837031275
  mocha-parallel-tests:runner-wrapped Emit suite event +1ms
Event occured: [suite]
[1477435036534] startSuite call (suites number: 2, id: 0.689213837031275
  mocha-parallel-tests:runner-wrapped Emit test event +1ms
Event occured: [test]
Event occured: [test]
[1477435036535] getCurrentSuite call (suites number: 2, id: 0.689213837031275
ADD FEATURE
ADD LABEL
[1477435036527] getCurrentSuite call (suites number: 0, id: 0.689213837031275
Cannot read property 'currentTest' of undefined
```

Despite the fact that console.log is happening in random order (why the hell?) it all means that allure.feature was called before "suite" events were fired.

**Comment by 1999** [comment]
Okay, I finally figured out what's the problem though I can't find a simple solution for now.

Test code (`it` callback) is executed immediately while mocha-allure-reporter is waiting for runner events (it adds suites when `suite` event is fired and removes when `suite end` occurs). The problem is that mocha-parallel-tests runner's events are throttled for proper `console.log`s and this is the reason no suite exists when `allure.reporter` is executed.

**Comment by 1999** [comment]
Just wanted to post this. I have finally implemented some kind of solution (#83) which works only for synchronous runner callbacks:
```javascript
// reporter file

// works
runner.on('pass', function (test) {
  console.log('Test passed: ' + test.file);
});

// doesn't work
runner.on('pass', function (test) {
  fs.writeFile('file.txt', 'utf-8', function () {
    console.log('Test passed: ' + test.file);
  });
});
```

The idea is to emit runner events as soon as they appear but intercept all standard streams (stderr & stdout) messages. When test file executing comes to an end (`end` event is fired), get all intercepted messages by test file and output them immediately.

However there are some problems with this approach:

1) Currently mocha-parallel-tests supports multiple describes inside one file. So one cannot distinguish between console.logs from one describe and another inside one file
2) Reporter reaction to runner events has no test file in its stack trace (this is how I detect test files: parse stack trace). This is the reason I implemented a **hack** method `Interceptor. prototype.setFileForEventTick` which sets test file just for one tick. This is dirty.
3) Some tests currently fail. This can probably be fixed.

**Comment by FelixZilber** [comment]
Anyone have a solution for executing in parallel with allure reporter?

**Comment by 1999** [comment]
This issue is fixed in master. The new release is coming in the next few days.

**Comment by 1999** [comment]
@FelixZilber @alexKazarin sorry for a long wait. This issue is fixed in mocha-parallel-tests@2.0.1.
